### PR TITLE
Using getopt for CLI arguments

### PR DESCRIPTION
**Pull request by nupplaphil**
Adding Arguments Class to Friendica and use them for daemon/worker.

Based on #5383

### CODE REVIEW

File: bin/worker.php [hunk 1]
```diff
@@ -41,14 +41,14 @@
 
 Addon::loadHooks();
 
-$spawn = (($_SERVER["argc"] == 2) && ($_SERVER["argv"][1] == "spawn"));
+$spawn = Argument::get('spawn',$_SERVER["argc"], false, 'boolean');
```
Comment (MrPetovan) [inline bin/worker.php]: `$_SERVER["argc"]` holds the argument count, not the argument array, please use `$_SERVER['argv']` instead.
Additionally, space is missing after a comma.

File: bin/worker.php [hunk 2]
```diff
@@ -41,14 +41,14 @@
...
 
 if ($spawn) {
 	Worker::spawnWorker();
 	killme();
 }
 
-$run_cron = (($_SERVER["argc"] <= 1) || ($_SERVER["argv"][1] != "no_cron"));
+$run_cron = Argument::get('cron',$_SERVER["argc"], false, 'boolean');
```
Comment (MrPetovan) [inline bin/worker.php]: `$_SERVER["argc"]` holds the argument count, not the argument array, please use `$_SERVER['argv']` instead.
Additionally, space is missing after a comma.

File: src/Core/Worker.php [hunk 3]
```diff
@@ -1010,13 +1010,11 @@ private static function runCron()
 	 */
 	public static function spawnWorker($do_cron = false)
 	{
-		$args = ["bin/worker.php"];
+		$command = "bin/worker.php";
```
Comment (MrPetovan) [inline src/Core/Worker.php]: Code standards: Please use single quotes by default.

File: src/Core/Worker.php [hunk 4]
```diff
@@ -1010,13 +1010,11 @@ private static function runCron()
...
 
-		if (!$do_cron) {
-			$args[] = "no_cron";
-		}
+		$args[] = [ "cron" => $do_cron ];
```
Comment (MrPetovan) [inline src/Core/Worker.php]: Code standards: Please use single quotes by default.

File: src/Util/Argument.php [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+	 * @param string $key     The search key
+	 * @param array  $args    The arguments to search
+	 * @param string $default The default value if no value was found (default empty)
+	 * @param string $type    The type of value it has to be (default null = all)
+	 *
+	 * @return string The value to the key
+	 *
+	 * @throws InternalServerErrorException Arguments are not valid
+	 */
+	public static function get($key, $args, $default = '', $type = null) {
```
Comment (MrPetovan) [inline src/Util/Argument.php]: Now that I'm looking at it in isolation, I feel like much is wrong with this class. If I understand clearly it's used to convert from command line arguments to a hash and vice-versa.

Therefore, I think it should be an actual object instantiated with the `$_SERVER['argv']` value by default or an array of arguments. You can even throw 

Suggested use case:
```php
$arguments = new CLIArguments(); // assuming $_SERVER['argv']
//or
$arguments = new CLIArguments($argv);

$arguments->get('key', $default, $type);
$arguments->set('newkey', $value);
// Returns a string of command line arguments
$args = $arguments->toString();

$command .= '/usr/bin/php ' . $args;
```

Additionally, you have to make sure that the returned command line argument string works with the current OS (see usage in your src/App.php:1134).

You can even throw `ArrayAccess` for good measure! See http://php.net/manual/en/class.arrayaccess.php

```php
class CLIArguments implements ArrayAccess {
    ...
}

$arguments = new CLIArguments(); // assuming $_SERVER['argv']
//or
$arguments = new CLIArguments($argv);

$value = $arguments['key'];
$arguments['newkey'] = $value;

// Returns a string of command line arguments
$args = $arguments->toString();

$command .= '/usr/bin/php ' . $args;
```

Comment (nupplaphil) [reply src/Util/Argument.php]: Thank you @MrPetovan !
That's why I want to split the big PR in different smaller PRs.

It looks much better than the static class :-)

**Comment by nupplaphil** [comment]
I think I over complicated things!

- There is already the `getopt()` function in PHP which does a great job for arguments.
- Passing the right arguments to `proc_run` doesn't justify a new class

File: src/Core/Worker.php [hunk 6]
```diff
@@ -1010,13 +1010,11 @@ private static function runCron()
 	 */
 	public static function spawnWorker($do_cron = false)
 	{
-		$args = ["bin/worker.php"];
+		$command = 'bin/worker.php';
 
-		if (!$do_cron) {
-			$args[] = "no_cron";
-		}
+		$args[] = [ 'cron' => $do_cron ];
```
Comment (MrPetovan) [inline src/Core/Worker.php]: The `[]` are superfluous.

**Comment by Quix0r** [comment]
@nupplaphil Don't worry! I think this can be resurrected. Maybe you can rename it to `Request` instead? Then console parameters can be fetched by `getopt()` and this class will also encapsulate `$_GET`, `$_POST` and `$_SESSION` in one single class.

**Comment by nupplaphil** [comment]
@Quix0r You're right! But I want to finish #5383 first. So I'd prefer to merge this PR, I'll finish the process PR and afterwards we can optimize things :-)
- removing `killme()`
- introducing a `Request` class

**Pull request merged**